### PR TITLE
Todo Item Description and View Modal Implemented

### LINKED ISSUE #421: Create TODO message field, modal pop-up, and rating
1. When creating or editing a TODO item, add a description field that can contain links (just like posts)
Screen Shot 2022-12-08 at 1 41 35 AM

2. When item item is clicked open a modal pop-up for viewing the item, containing all fields (as seen below)
 - title (at top)
 - description (links should be hyperlinked, just like post descriptions) -- just like posts
 - type (value or "None"); (e.g. Type: Social-emotional learning)
 - name (i.e., either the personal name or the group name); (can be just below the title)
 - date (can include a "schedule" icon, date, time, and the status: Completed (Green), Pending (Yellow), or Missed (Red)-- just like CK Monitor - TODO monitor 
Screen Shot 2022-12-21 at 11 41 09 AM

3. In addition, we will no longer support marking multiple tasks as complete at the same time. Instead, in the modal pop-up for each item there will be a mandatory completion rating and a complete button. 

 - The "Completion Quality" selector will be defaulted to "N/A" and have additional values "Very unsatisfied", "Unsatisfied", "Neutral", "Satisfied", and "Very satisfied"
 - Users must select an option other than "N/A" in order to click the "Mark as Completed" button

Screen Shot 2022-12-21 at 11 58 39 AM

### LINKED ISSUE #452: Rename TODO Types
1. Let's rename the following types:

 - Cognitive strategies --> Content learning 
 - Social-emotional learning supports --> Social-emotional learning
 - Behavioural engagement strategies --> Classroom engagement

2. Also, can we update the tooltip info to:
Content learning (i.e., learning strategies for memorization, recall, or application of knowledge)

Social-emotional learning (i.e., supports for developing your emotional awareness, stress management, motivation, math identity, relationships, or critical/creative thinking)

Classroom engagement (i.e., practices for improving learning engagement, e.g., note-taking, asking questions, designing/explaining/suggesting ideas, or generating/sharing ideas with classmates)

P.S. Don't worry about clobbering existing info, I have a record of their TODOs so I can reset the type values

### PR DESCRIPTION
**Pull request by LunarFang416**
## Details

- Todo Item Description field now added, which also supports links and formats accordingly
- Uncompleted Items can now only be marked as completed through the individual todo item view modal, where you must give a rating to complete a task
- Already completed tasks will show the completion quality in the todo item view modal
- Todo Item types also renamed as specified

Closes #421 , Closes #452

### CODE REVIEW

**Comment by JoelWiebe** [comment]
@LunarFang416 Wow, you made quick work of this! Impressive. 

For some reason, I'm getting an error each time I try to create a new TODO item. Are you getting this as well?
Screen Shot 2022-12-21 at 9 54 59 PM

**Comment by JoelWiebe** [comment]
P.S. This should overwrite the old item Types but still leave the other fields of old items, right? For some reason my previously created TODO items don't seem to be showing up either right now.

**Comment by LunarFang416** [comment]
@JoelWiebe I'm unable to recreate the error that you are getting for some reason. I was able to create todo items in a new board with no errors. Thats strange.

For the old todo items, they are causing errors since they don't contain the new added fields: description and completion quality.

**Comment by JoelWiebe** [comment]
I tried on a new board using a teacher account, and somehow I'm still getting this error below when I initially load the TODO List modal, then the error in my previous message when creating a new item.
Screen Shot 2022-12-21 at 11 02 38 PM

Are all of your changes in this commit? Is there anything else I could check?

**Comment by JoelWiebe** [comment]
@LunarFang416 Thanks for figuring that out! Clearly I'm not sleeping enough again, haha. Yes, I should have tried creating a new project not a new board. Is there a way that this could fail more gracefully or that we could easily add the missing fields for existing projects (as we have an existing project that students will continue to use and build onto their TODOs)?

**Comment by LunarFang416** [comment]
@JoelWiebe I made the changes so that older todo lists wont cause problems. Please let me know if it works smoothly on your end.

**Comment by JoelWiebe** [comment]
Brilliant! Great work

Tested:

 - TODOs can be opened in modal pop-up
 - Authors are listed as either the individual or the group name
 - TODO items can be created and edited with a description field that can include links
 - Items require a rating before being completed

Suggestions:

1. Can we add hyperlink styling to the task title for the personal and group tabs (so that students know these rows are clickable)

(see current task titles below)
Screen Shot 2022-12-28 at 4 57 15 PM

2. Can we (a) re-order and (b) rename the select drop-down for marking as complete (e.g., N/A, Poor, Low, Moderate, Good, Excellent -- in that order)? 

(currently select drop-down list is not in the ordinal order)
Screen Shot 2022-12-28 at 3 34 05 PM

3. Can we re-order the select drop-down for TODO types in the following order? (e.g., Content learning, Social-emotional learning, Classroom engagement, and Assigned class work) -- NOTE: I re-named that last one

(the current select drop-down list order is displayed below, can we replace this order with the above ordering)
Screen Shot 2022-12-28 at 4 50 09 PM

**Comment by JoelWiebe** [comment]
Awesome work. Everything looks great!

Sorry, I just remembered one small item: If an item is marked as done with the "Incomplete" status, could we skip the confetti and music?

**Review by JoelWiebe — APPROVED** [review]
Brilliant work!

**Comment by JoelWiebe** [comment]
This one looks good to me. @Ammar-T can you give this one a quick review when you get a chance?

File: frontend/src/app/services/todoItem.service.ts [hunk 1]
```diff
@@ -57,79 +57,79 @@ export class TodoItemService {
...
+  //         this.update(todoItems[i].todoItemID, {
+  //           notifications: [
+  //             ...todoItems[i].notifications,
+  //             TODOITEM_NOTIFICATION.SECONDARY,
+  //           ],
+  //         });
+  //       }
+  //     }
+  //   }
+  // }
```
Comment (DCH25) [inline frontend/src/app/services/todoItem.service.ts:134]: Remove if we're not using it anywhere

File: frontend/src/app/components/todo-list-modal/todo-list-modal.component.scss [hunk 2]
```diff
@@ -2,10 +2,14 @@
...
+.todoItemRow:hover{
+  cursor: pointer;
+  background: #f5f5f5;
+}
+
 .mat-column-task-title {
-  padding-left: 16px;
   font-size: 16px;
-  margin-right: 16px;
+  padding:0 12px;
```
Comment (DCH25) [inline frontend/src/app/components/todo-list-modal/todo-list-modal.component.scss:12]: nitpick: add spacing

File: frontend/src/app/components/todo-list-modal/todo-list-modal.component.scss [hunk 3]
```diff
@@ -14,6 +18,11 @@
   padding-right: 16px;
 }
 
+.mat-column-quality{
+  font-size: 16px;
+  padding:0 12px;
```
Comment (DCH25) [inline frontend/src/app/components/todo-list-modal/todo-list-modal.component.scss:23]: add spacing

File: frontend/src/app/components/todo-item-card-modal/todo-item-card-modal.component.html [hunk 4]
```diff
@@ -0,0 +1,80 @@
+<div>
+  <mat-card class="details">
+    <mat-card-header>
+      <div mat-card-avatar class="avatar"></div>
+      <mat-card-title>{{ taskTitle }}</mat-card-title>
+      <span class="fill-remaining-space"></span>
+      <mat-card-subtitle>{{assignee}}</mat-card-subtitle>
```
Comment (DCH25) [inline frontend/src/app/components/todo-item-card-modal/todo-item-card-modal.component.html:7]: spacing

**Review by DCH25 — CHANGES_REQUESTED** [review]
small style changes o/w LGTM

**Pull request merged**